dev-tools/ls-apis/src/system_apis.rs

```diff
@@ -0,0 +1,629 @@
+            .collect();
+
+        // Now walk through the deployment units, walk through each one's server
+        // packages, walk through each one of the clients used by those, and
+        // create a corresponding edge.
+        for deployment_unit in self.deployment_units() {
+            let server_components =
+                self.deployment_unit_servers(deployment_unit).unwrap();
+            let my_node = nodes.get(deployment_unit).unwrap();
+            for server_pkg in server_components {
+                for (client_pkg, _) in
+                    self.component_apis_consumed(server_pkg, filter)?
+                {
+                    let other_component =
+                        self.api_producer(client_pkg).unwrap();
+                    let other_unit = self
+                        .server_component_units
+                        .get(other_component)
+                        .unwrap();
+                    let other_node = nodes.get(other_unit).unwrap();
+                    graph.add_edge(*my_node, *other_node, client_pkg.clone());
+                }
+            }
+        }
+
+        Ok(Dot::new(&graph).to_string())
+    }
+
+    /// Returns a string that can be passed to `dot(1)` to render a graph of
+    /// API dependencies among server components
+    pub fn dot_by_server_component(
+        &self,
+        filter: ApiDependencyFilter,
+    ) -> Result<String> {
+        let mut graph = petgraph::graph::Graph::new();
+        let nodes: BTreeMap<_, _> = self
+            .server_component_units
+            .keys()
+            .map(|server_component| {
+                (server_component.clone(), graph.add_node(server_component))
+            })
+            .collect();
+
+        // Now walk through the server components, walk through each one of the
+        // clients used by those, and create a corresponding edge.
+        for server_component in self.apis_consumed.keys() {
+            // unwrap(): we created a node for each server component above.
+            let my_node = nodes.get(server_component).unwrap();
+            let consumed_apis =
+                self.component_apis_consumed(server_component, filter)?;
+            for (client_pkg, _) in consumed_apis {
+                let other_component = self.api_producer(client_pkg).unwrap();
+                let other_node = nodes.get(other_component).unwrap();
+                graph.add_edge(*my_node, *other_node, client_pkg.clone());
+            }
+        }
+
+        Ok(Dot::new(&graph).to_string())
+    }
+}
+
+/// Helper for building structures to index which deployment units contain which
+/// server components and what APIs those components expose
+///
+/// See `SystemApis::load()` for how this is used.
+struct ServerComponentsTracker<'a> {
+    // inputs
+    known_server_packages: &'a BTreeMap<ServerPackageName, &'a ApiMetadata>,
+
+    // outputs (structures that we're building up)
+    errors: Vec<anyhow::Error>,
```

Comment:
Yeah, that's interesting.  But I think this would be more unwieldy in this case.  This is kind of like a builder with a bunch of different methods that could produce errors.  I think it'd be much more annoying to have to pass `errors` to all of the methods.